applications/tari_console_wallet/src/ui/state/debouncer.rs

```diff
@@ -0,0 +1,139 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::ui::state::AppStateInner;
+use log::*;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tari_wallet::output_manager_service::handle::OutputManagerHandle;
+use tokio::{
+    sync::{broadcast, RwLock},
+    time,
+};
+
+const LOG_TARGET: &str = "wallet::console_wallet::debouncer";
+
+#[derive(Clone)]
+pub(crate) struct BalanceEnquiryDebouncer {
+    app_state_inner: Arc<RwLock<AppStateInner>>,
+    output_manager_service: OutputManagerHandle,
+    balance_enquiry_cooldown_period: Duration,
+    tx: broadcast::Sender<()>,
+}
+
+impl BalanceEnquiryDebouncer {
+    pub fn new(
+        app_state_inner: Arc<RwLock<AppStateInner>>,
+        balance_enquiry_cooldown_period: Duration,
+        output_manager_service: OutputManagerHandle,
+    ) -> Self {
+        // This channel must only be size 1; the debouncer will ensure that the balance is updated timeously
+        let (tx, _) = broadcast::channel(1);
+        Self {
+            app_state_inner,
+            output_manager_service,
+            balance_enquiry_cooldown_period,
+            tx,
+        }
+    }
+
+    pub async fn run(mut self) {
+        let balance_enquiry_events = &mut self.tx.subscribe();
+        let mut shutdown_signal = self.app_state_inner.read().await.get_shutdown_signal();
+        let delay = time::sleep(self.balance_enquiry_cooldown_period);
+        tokio::pin!(delay);
+
+        debug!(target: LOG_TARGET, "Balance enquiry debouncer starting");
+        if let Ok(balance) = self.output_manager_service.get_balance().await {
+            trace!(
+                target: LOG_TARGET,
+                "Initial balance: available {}, incoming {}, outgoing {}",
+                balance.available_balance,
+                balance.pending_incoming_balance,
+                balance.pending_outgoing_balance
+            );
+            let mut inner = self.app_state_inner.write().await;
+            if let Err(e) = inner.refresh_balance(balance).await {
+                warn!(target: LOG_TARGET, "Error refresh app_state: {}", e);
+            }
+        }
+        loop {
+            tokio::select! {
+                _ = &mut delay => {
```

Comment:
If the desired behaviour is to cache the result for multiple calls within the cooldown period, this is not being achieved.

What will happen here is that many calls to this method will get slower and slower, blocking the callers.

Wouldn't it be better to change it to something like 

```

select! {
  balance_enquiry_events.recv() =>  {
   if  Instant::now() < cache_until_time { return cached_value }
  // get value
   cache_until_time = Instant::now() + self.balance_enquiry_cooldown_period;
   return value;
 }
  
}
```

Reply:
> If the desired behaviour is to cache the result for multiple calls within the cooldown period, this is not being achieved.

The desired behaviour is to postpone all calls to update the balance for a set cooldown period of time till the very end of the period (and of course then only perform one calculation) and not update it at the beginning of the period and use the cached value. This will ensure that the last request to update the balance will always be honoured if no more requests arrive after the current cooldown period. If the cooldown period is made much bigger for example increased to 60 s for large busy wallets, then not honouring the last balance update request in the cooldown period is very much noticeable in the TUI with displaying the incorrect balance.

> What will happen here is that many calls to this method will get slower and slower, blocking the callers.

If 1000 requests are sent to the debouncer (`broadcast`) channel in 1 s then only the last one will be remembered as it has size 1; all prior requests will be dropped. The select loop will only check for a new update request after the delay event, and the channel receive future will timeout if there is none. Thus no blocking calls.